plugins/alembic/module/vtkF3DAlembicReader.cxx

```diff
@@ -20,51 +22,338 @@
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f originalUV =
+              originalData.Attributes.at("uv")[originalData.Indices[i][j].y];
+            uvV3F.emplace_back(originalUV);
+            duplicatedData.Indices[i][j].y = uvRunningIndex;
+            uvRunningIndex++;
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("uv", uvV3F));
+      }
+
+      // Normal
+      if (haveN)
+      {
+        V3fContainer nV3F;
+        int nRunningIndex = 0;
+
+        for (size_t i = 0; i < faceCount; i++)
+        {
+          auto thisFaceVertexCount = originalData.Indices[i].size();
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f original_N =
+              originalData.Attributes.at("N")[originalData.Indices[i][j].z];
+            nV3F.emplace_back(original_N);
+            duplicatedData.Indices[i][j].z = nRunningIndex;
+            nRunningIndex++;
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("N", nV3F));
+      }
+    }
+    else
+    {
+      duplicatedData = originalData;
+    }
+  }
+
+  void FillPolyData(const IntermediateGeometry& data, vtkPolyData* polydata)
+  {
+    // Create 10 points.
+    vtkNew<vtkPoints> points;
+    vtkNew<vtkCellArray> cells;
+
+    auto pMapIter = data.Attributes.find("P");
+    if (pMapIter == data.Attributes.end())
+    {
+      // TODO throw an error instead and catch it above ?
+      vtkWarningWithObjectMacro(nullptr, "A data could not be read, leaving empty");
+      return;
+    }
+
+    // Note : uv and N are optional
+    auto uvMapIter = data.Attributes.find("uv");
+    auto nMapIter = data.Attributes.find("N");
+    bool haveUV = uvMapIter != data.Attributes.end();
+    bool haveN = nMapIter != data.Attributes.end();
+    for (auto& pIter : pMapIter->second)
+    {
+      points->InsertNextPoint(pIter.x, pIter.y, pIter.z);
+    }
+    polydata->SetPoints(points);
+
+    for (auto& faceIndicesIter : data.Indices)
+    {
+      vtkNew<vtkPolygon> polygon;
+      for (auto& vertexIndicesIter : faceIndicesIter)
+      {
+        polygon->GetPointIds()->InsertNextId(vertexIndicesIter.x);
+      }
+      cells->InsertNextCell(polygon);
+    }
+    polydata->SetPolys(cells);
+    vtkDataSetAttributes* pointAttributes = polydata->GetAttributes(vtkDataSet::POINT);
+    vtkDataSetAttributes* cellAttributes = polydata->GetAttributes(vtkDataSet::CELL);
+
+    assert(pointAttributes != nullptr);
+    assert(cellAttributes != nullptr);
+    if (haveN)
+    {
+      vtkNew<vtkFloatArray> normals;
+      normals->SetNumberOfComponents(3);
+      for (auto& N : nMapIter->second)
+      {
+        normals->InsertNextTuple3(N.x, N.y, N.z);
+      }
+      if (data.nIsFaceVarying)
+      {
+        cellAttributes->SetNormals(normals);
+      }
```

Comment:
do you have nFaceVarying data to test this with @nyue ?

Reply:
[tetrahedron_facevarying_N.zip](https://github.com/f3d-app/f3d/files/13398760/tetrahedron_facevarying_N.zip)

Reply:
I'm afraid opening above file with F3D does not reach this code. Is that expected ?